src/util/Histogram.h

```diff
@@ -0,0 +1,163 @@
+        double const val;       ///< value of the entry.
+    };
+
+    /// Constructor
+    /// @param label - Name for the histogram
+    /// @param bucketVals - a vector indicating the maximum value for each bucket.
+    /// @param maxSize - maximum number of entries to keep. Must be greater than 0.
+    /// @param maxAge - maximum age in milliseconds, such that entries older than
+    ///                       it will be removed from the histogram. Default is 1 hour.
+    Histogram(std::string const& label, std::vector<double> const& bucketVals,
+              std::chrono::milliseconds maxAge, size_t maxSize = 1000);
+    Histogram() = delete;
+    Histogram(Histogram const&) = delete;
+    Histogram& operator=(Histogram const&) = delete;
+
+    /// Add a time and value to the histogram.
+    std::string addEntry(TIMEPOINT stamp, double val, std::string const& note = "");
+
+    /// Add a time entry using CLOCK::now() as the time stamp.
+    std::string addEntry(double val, std::string const& note = "");
+
+    void setMaxSize(size_t maxSize);                   ///< Set the maximum number of entries to keep.
+    void setMaxAge(std::chrono::milliseconds maxAge);  ///< Set the maximum age for entries.
+
+    /// Return the count for bucket `index`, where index=0 is the bucket for the smallest values.
+    int getBucketCount(size_t index) const;
+
+    /// Return the _maxVal for the bucket at `index`.
+    double getBucketMaxVal(size_t index) const;
+
+    double getAvg() const;                        ///< Return the average value of all current entries.
+    double getTotal() const;                      ///< Return the total value of all entries.
+    double getSize() const;                       ///< Return the number of entries.
+    std::chrono::milliseconds getMaxAge() const;  ///< Return the maximum age allowed.
+    size_t getMaxSize() const;                    ///< Return the maximum number of entries allowed.
+    void checkEntries();                          ///< Remove old entries or entries beyond maxSize.
+    std::string getString(std::string const& note = "");  ///< @return a log worthy version of the histogram.
+
+    nlohmann::json getJson() const;  ///< Return a json version of this object.
+    std::string getJsonStr() const;  ///< Return the json string for this instance.
+
+private:
+    /// _mtx must be locked, return the average value of all current entries.
+    double _getAvg() const;
+
+    /// _mtx must be locked, returns a log worthy string.
+    std::string _getString(std::string const& note);
+
+    /// _mtx must be locked, Change the counts of the appropriate bucket by `incr`.
+    /// Normal values of `incr` should be -1 or 1.
+    void _changeCountsBy(double val, int incr);
+
+    /// Remove old entries, _mtx must be held when calling.
+    void _checkEntries();
+
+    std::string _label;
+    mutable std::mutex _mtx;
+    std::vector<Bucket> _buckets;
+    int64_t _overMaxCount = 0;
```

Comment:
Why the signed integer type is here? does the Histogram design allow negative counters?

Reply:
If there's a logic error with subtraction, massive positive values will be much more harmful than small negative numbers. It will also be nearly impossible to reach the positive maximum value of an int64_t.